### ISSUE TITLE
[Feature]:SequenceAccuracy Metric for PyTorch Ignite

### ISSUE THREAD

**Issue by aaishwarymishra** [pytorch/ignite#3593]
## 🚀 Feature

I think  the addition of a ```SequenceAccuracy``` metric (or an enhancement to the existing Accuracy metric) to natively support sequence-to-sequence outputs common in NLP and Transformer-based models will be good.

## Motivation:

When working with sequence models, the model output is typically of shape ```(batch, sequence_length, num_classes)``` and the target is ```(batch, sequence_length)```. Currently, the standard ignite.metrics.Accuracy expects inputs to be already flattened or simple 1D/2D tensors.

I have to write a custom output_transform for every single project to flatten the tensors and create masking.

Current work around:
```
def get_metrics(loss_fn, pad_idx=0):
    def accuracy_transform(output):
        preds, y = output

        preds_flat = preds.reshape(-1, preds.size(-1))
        y_flat = y.reshape(-1)

        mask = y_flat != pad_idx

        preds_masked = preds_flat[mask]
        y_masked = y_flat[mask]

        return preds_masked, y_masked

    return {
        "accuracy": metrics.Accuracy(output_transform=accuracy_transform),
        "loss": metrics.Loss(loss_fn),
    }

```
As ```Accuracy``` is one of the most common nlp metric I think offering a more plug and play option will be godd.

We can create a new metric ```SequenceAccuracy``` which can be a wrapper around ```Accuracy``` and perform masking, alternatively we can modify the ```Accuracy``` itself to support sequences and masking.

**Comment by aaishwarymishra** [comment]
Well I think making a separate metric will be better. I can work on this if you think its a good addition and after api discussion.

**Comment by vfdev-5** [comment]
@aaishwarymishra actually, I wonder if we could simply provide the output transform for sequences as you put in the workaround code, instead of creating a new class metric. For example, something like:
```python
from ignite.metrics import Accuracy

acc = Accuracy(output_transform=Accuracy.get_sequence_transform(pad_index=0))
# optionally we can still provide the output transform applied before the sequence transform:
# acc = Accuracy(output_transform=Accuracy.get_sequence_transform(pad_index=0, output_transform=lambda out: out))
```

**Comment by Prathamesh8989** [comment]
Hi @vfdev-5 and @aaishwarymishra,

I like the idea of providing `Accuracy.get_sequence_transform(pad_index=...)` instead of introducing a new metric class — it keeps the API clean and avoids duplication.

A couple of questions for clarification before implementation:

1. Should the transform also support an `ignore_index` argument similar to `nn.CrossEntropyLoss`?
2. Do we want to handle both `(batch, seq_len, num_classes)` logits and already-argmaxed predictions?
3. Should masking support variable-length sequences coming from packed sequences, or only simple padding-based masking?

If this direction sounds good, I’d be happy to take a shot at implementing `get_sequence_transform`.

Looking forward to your thoughts!

**Comment by aaishwarymishra** [comment]
I was originally planning on working on it myself, But I am kinda busy with college and stuff.

> 1. Should the transform also support an `ignore_index` argument similar to `nn.CrossEntropyLoss`?

Are you talking about `ignore_index' like `pad_idx` in my example transform? If yes than yeah also it should be named `pad_idx` or something similar so user can know this takes padding id. In my example we are only ignoring one index value but we should add support for multi-index ignoring.
> 3. Do we want to handle both `(batch, seq_len, num_classes)` logits and already-argmaxed predictions?

yes, Base Accuracy metric already handles `(batch, num_classes)`  and `(batch, 1)`  the main job of the transform is to convert the `(batch, seq_len, num_classes)` to `(batch, num_classes)` or `(batch,1)` accordinly.
> 3. Should masking support variable-length sequences coming from packed sequences, or only simple padding-based masking?

I am not sure what you mean by this.

**Comment by vfdev-5** [comment]
> Should masking support variable-length sequences coming from packed sequences, or only simple padding-based masking?

It's true that targets (y) can be a packed sequence. Supporting that may make sense as well.

<img src="https://cdn-uploads.huggingface.co/production/uploads/6041ff7ff84ebe399f1c85ea/wfZM7YcPyvS0qYnwhdqic.png"/>

Source: https://huggingface.co/blog/sirluk/llm-sequence-packing

```python
# y target contains 3 sequences:
    y = [[1, 2, 3, 4, 9, 1, 5, 9, 4, 5, 6, 9, <pad>, <pad>, <pad>]]
y_pos = [[0, 1, 2, 3, 4, 0, 1, 2, 0, 1, 2, 3, <pad>, <pad>, <pad>]]
```

**Comment by aaishwarymishra** [comment]
Do you think adding support for sequence reduction be good?

Example:
```python
x = [[1,1,1,1,1],[1,1,1,1,1]]
y = [[1,1,1,1,1],[1,1,1,1,0]]

# without sequence reduction accuracy will be 0.9 as one token doesn't match

# with sequence reduction accuracy will be 0.5 because only 1 sequenced matched out of 2
```

**Comment by Niya3-Navya** [comment]
Hi, I’m interested in contributing to this issue as part of my GSoC preparation. 
I have experience with ML projects (LSTM, NLP) and would like to work on adding tutorial notebooks. 
Could you suggest a specific example or topic to start with?

**Comment by aaishwarymishra** [comment]
@Niya3-Navya  you can look at issue #3547

**Comment by kyleclai** [comment]
Hi, I’m interested in contributing to this issue as part of my open-source contribution program.
I have experience with ML projects (image classification pipelines, custom metrics) and would like to work on the sequence reduction mode.
I plan to get started next week — could you suggest a specific example/topic to begin with?

**Comment by aaishwarymishra** [comment]
Hi, @kyleclai we have just merged the pr for the sequence accuracy, but we also want to support packing with it.

If you are interested you can work on supporting that with sequence accuracy :)

vfdev-5 has talked about it in the issue above , but you can read more about it in : https://huggingface.co/blog/sirluk/llm-sequence-packing
